### PR TITLE
Add SqueezeTransform: scroll-specific compression augmentation (#201)

### PR DESCRIPTION
**Pull request by pscamillo**
# Add `SqueezeTransform`: scroll-specific compression augmentation (#201)

Implements the **Squeezing/Pulling** augmentation requested in #201 — "create
accurate compressed regions from normal data."

## What

`batchgeneratorsv2/transforms/spatial/squeeze.py` — a new `SqueezeTransform`
(subclass of `BasicTransform`), drop-in alongside `SpatialTransform`.

## How it works

For a chosen axis it builds a non-negative compression profile, takes its
normalised cumulative integral as a resampling map (strictly monotonic ⇒ no
folding), and `grid_sample`s with it: layers bunch where the profile peaks
(compression), flanks stretch. Unlike elastic deformation (isotropic/random
blurred offsets), this is the structured, anisotropic warp that real scroll
compression actually produces. The displacement is a single 1-D `cumsum`
broadcast over the patch, built and applied **entirely on the input tensor's
device** — no CPU/NumPy FFT round-trip.

## Why it's useful

Compressed regions are a documented weak spot of the surface/fiber/ink models
(#191; discussion #39). This augmentation generates more of that under-represented
geometry at train time.

## Evidence

- **Ablation** (controlled synthetic, same model/seed/data, only the augmentation
  differs): **+0.0395 Dice (+4.7%)** on held-out compressed geometry the model
  never saw in training.
- **Speed** (160³, image+seg): **4.8× faster** than the elastic path on CPU
  (106 ms vs 514 ms/sample) and **~4 ms/sample on GPU** — the elastic path is
  CPU-bound (NumPy FFT + CPU grid) and does not run on GPU.

## Details

- API: standard `get_parameters` / `apply`; image bilinear, segmentation nearest;
  the same displacement field is applied to image and labels (geometric
  consistency).
- Works for 2D `(C,X,Y)` and 3D `(C,X,Y,Z)` tensors.
- Includes a `__main__` self-test (identity-when-disabled, shape/label
  preservation, real compression, per-sample timing).
- A reproducible ablation + benchmark + visualization script is available
  (not part of the package; provided separately).

## Scope / next steps

The ablation here is a controlled synthetic study that isolates the mechanism; an
ablation inside the full nnU-Net training on real scroll data is the natural
follow-up. A fiber-direction-aware variant (using the fiber predictions from #203)
would orient the squeeze along the local sheet normal.

Addresses the Squeezing/Pulling item of #201.

---

*Developed with AI pair-programming assistance (Claude), reviewed and run by the
author; noted for transparency.*
after_image
after_mask

### CODE REVIEW

**Comment by bruniss** [comment]
thanks for the PR! do you have an example of this transform on real data ? it would address a need in our training pipeline but synthetic examples dont always translate well

**Comment by pscamillo** [comment]
Thanks! Agreed synthetic ablations don't always transfer, so here it is on a real annotated cube from the data — Scroll 1 (PHercParis4) instance-label cube `02256_02512_04816` (`_volume.nrrd` + `_mask.nrrd`, 256³, 16 sheet instances), loaded via the `vesuvius` library.

Top row: original CT + instance label. Bottom: after `SqueezeTransform` along the across-sheet axis (picked automatically as the axis with the highest label-transition density). The papyrus sheets bunch into a realistic compressed band and the instance label warps with the exact same displacement field (nearest-neighbour for labels), so it stays voxel-accurate and drop-in for the patch pipeline.

Your question also caught a rough edge — I dropped a stray mean-subtraction in the displacement so the warp now stays fully in-bounds (no zero-padding at the patch border, endpoints pinned). Pushed to the branch (3590d9e).

Happy to take it further and run a real-data ablation in your setup: if you point me at the canonical training config (the nnU-Net-style surface model), I'll train with/without the aug and report Dice on held-out compressed regions.
real_squeeze_demo

**Comment by giorgioangel** [comment]
@codex review

File: segmentation/models/batchgeneratorsv2/batchgeneratorsv2/transforms/spatial/squeeze.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        for _ in range(n_bands):
+            # keep band centers away from the very edges so the squeeze is visible
+            center = float(torch.empty(1, device=device).uniform_(0.15 * L, 0.85 * L).item())
+            sigma_px = max(1.0, float(sample_scalar(self.sigma, image=img, axis=axis)) * L)
+            strength = max(0.0, float(sample_scalar(self.strength, image=img, axis=axis)))
+            c = c + strength * torch.exp(-((xs - center) ** 2) / (2.0 * sigma_px ** 2))
+
+        # normalised cumulative integral -> strictly monotonic resampling map
+        m = torch.cumsum(c, dim=0)
+        m = (m - m[0]) / (m[-1] - m[0]) * (L - 1)
```
[the remark this answers is not in the document]
Comment (pscamillo) [reply segmentation/models/batchgeneratorsv2/batchgeneratorsv2/transforms/spatial/squeeze.py:158]: Fixed in c5d2eb0 — get_parameters now filters candidate axes to length >= 2 and returns a no-op for a singleton spatial dim (e.g. (C, 1, H, W)), so the cumulative map can't divide by zero / produce a NaN. The same commit also aligns the scalar API (dim= instead of axis=). Added self-test coverage for the singleton path.
